backend/Rhyolite/Backend/HaveIBeenPwned.hs

```diff
@@ -0,0 +1,148 @@
+
+-- Module extracted from focus. Got rid of all the stuff that is not
+-- needed and changed parts to use Rhyolite instead of Focus stuff (e.g.
+-- logging):
+module Rhyolite.Backend.HaveIBeenPwned where
+
+import "cryptohash" Crypto.Hash -- or maybe i wanted cryptonite?
+import Control.Monad.Reader
+import Control.Monad.Logger
+import Control.Exception
+import Data.ByteString (ByteString)
+import qualified Data.ByteString.Lazy as LBS
+import Data.Text(Text)
+import Data.Text.Encoding(encodeUtf8)
+import qualified Data.Text.Lazy.Encoding
+import Network.HTTP.Client
+import Network.HTTP.Client.TLS
+import Network.HTTP.Types.Status(Status(..))
+import System.IO
+import qualified Data.Text as T
+import qualified Data.Text.Lazy as LT
+import Data.Default (def)
+
+import Rhyolite.Backend.Logging (withLogging, LoggingConfig (..), RhyoliteLogAppender (..))
+
+
+data HaveIBeenPwnedError
+  = HaveIBeenPwnedError_BadResponse Status
+  | HaveIBeenPwnedError_NetworkError HttpException
+  deriving (Show) -- no eq instance for HttpException
+
+data HaveIBeenPwnedConfig = HaveIBeenPwnedConfig
+  { _haveIBeenPwnedConfig_manager :: Manager
+  , _haveIBeenPwnedConfig_apihost :: Text
+  }
+
+data HaveIBeenPwnedResult
+  = HaveIBeenPwnedResult_Undisclosed
+  | HaveIBeenPwnedResult_Disclosed Int
+  | HaveIBeenPwnedResult_ApiError
+  deriving (Eq, Ord, Show)
+
+class Monad m => MonadPwned m where
+  -- | returns the number of disclosures the supplied password has been seen in.
+  --
+  -- if this is not zero, do not use the supplied password, it is known to hackers.
+  -- if it is *is* zero, it might still not be safe, only that if it is
+  -- compramised, that is not yet known
+  --
+  -- https://haveibeenpwned.com/API/v2#SearchingPwnedPasswordsByRange
+  haveIBeenPwned :: Text -> m HaveIBeenPwnedResult
+
+newtype PwnedT m a = PwnedT { unPwnedT :: ReaderT HaveIBeenPwnedConfig m a }
+  deriving (Functor, Applicative, Monad , MonadIO, MonadLogger
+    , MonadTrans
+    )
+
+runPwnedT :: PwnedT m a -> HaveIBeenPwnedConfig -> m a
+runPwnedT (PwnedT (ReaderT f)) = f
+
+mapPwnedT :: (m a -> n b) -> PwnedT m a -> PwnedT n b
+mapPwnedT f = PwnedT . mapReaderT f . unPwnedT
+
+instance MonadReader r m => MonadReader r (PwnedT m) where
+  ask = lift ask
+  local = mapPwnedT . local
+  reader = lift . reader
+
+instance (MonadLogger m, MonadIO m) => MonadPwned (PwnedT m) where
+ haveIBeenPwned password = do
+  let (pfx, rest) = passwdDigest password
+  cfg <- PwnedT ask
+  let request = parseRequest_ $ T.unpack $ T.concat [_haveIBeenPwnedConfig_apihost cfg, "/", pfx]
+  result' <- liftIO $ try $ httpLbs request (_haveIBeenPwnedConfig_manager cfg)
+  case result' of
+    Left err -> do
+      $(logError) $ T.pack $ show @ HttpException $ err
+      return HaveIBeenPwnedResult_ApiError
+    Right result -> case responseStatus result of
+      Status 200 _ -> return $ case parseHIBPResponse (responseBody result) rest of
+        0 -> HaveIBeenPwnedResult_Undisclosed
```

Comment:
`HaveIBeenPwnedResult_Disclosed 0` is weird. Should we make this into a smart constructor and only export that?

Reply:
Good point.